libraries/apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/interceptor/RetryOnErrorInterceptor.kt

```diff
@@ -0,0 +1,71 @@
+package com.apollographql.apollo3.interceptor
+
+import com.apollographql.apollo3.annotations.ApolloExperimental
+import com.apollographql.apollo3.api.ApolloRequest
+import com.apollographql.apollo3.api.ApolloResponse
+import com.apollographql.apollo3.api.Operation
+import com.apollographql.apollo3.api.Subscription
+import com.apollographql.apollo3.exception.ApolloException
+import com.apollographql.apollo3.exception.SubscriptionOperationException
+import com.benasher44.uuid.uuid4
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.emitAll
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.retryWhen
+import kotlin.math.pow
+import kotlin.time.Duration.Companion.seconds
+
+@ApolloExperimental
+class RetryOnErrorInterceptor(shouldIntercept: ((ApolloRequest<*>) -> Boolean)? = null): ApolloInterceptor {
+  private val shouldIntercept: ((ApolloRequest<*>) -> Boolean) = shouldIntercept ?: {
+    when (it.retryNetworkErrors) {
+      true -> true
+      false -> false
+      null -> it.operation is Subscription
+    }
+  }
+
+  override fun <D : Operation.Data> intercept(request: ApolloRequest<D>, chain: ApolloInterceptorChain): Flow<ApolloResponse<D>> {
+    if (!shouldIntercept(request)) {
+      return chain.proceed(request)
+    }
+
+    var first = true
+    var attempt = 0
+    return flow {
+      val actualRequest = if (first) {
+        first = false
+        request
+      } else {
+        request.newBuilder().requestUuid(uuid4()).build()
+      }
+      emitAll(chain.proceed(actualRequest))
+    }.onEach {
+      if (it.exception != null && it.exception!!.isTerminalAndRecoverable()) {
+        throw RetryException
+      } else {
+        attempt = 0
+      }
+    }.retryWhen { cause, _ ->
+      attempt++
+
+      delay(2.0.pow(attempt).seconds)
+      cause is RetryException
+    }
+  }
+}
+
+
+private fun ApolloException.isTerminalAndRecoverable(): Boolean {
```

Comment:
Would it make sense to have this logic as a configurable lambda on `ApolloResponse`? So one could check for `exception` but also `errors` or even `data`.

Reply:
I have bikeshedded the API a bit:

* `retryNetworkErrors` -> `retryOnError` because it's shorter and pronounces better IMO
* There is now a required `defaultRetryOnError` in `RetryOnErrorInterceptor` so as to not break the symmetry of `RetryOnErrorInterceptor`

Overall I'm not sure yet how much retry logic belongs to user land vs something that we can provide with a bunch of parameters (the lambda you're proposing, + exponential backoff + Android Connectivity manager, maybe auth, etc...). My hunch is that it's certainly something that belongs to user land but maybe after some usage we can identify a minimal set of params that would serve the most use cases.

tldr; I would make `setRetryOnErrorInterceptor` API stable soon but wait a lot more for `RetryOnErrorInterceptor`

Reply:
I ended up changing thing again [here](https://github.com/apollographql/apollo-kotlin/pull/5690/files#diff-aa6dd8789c20afff5b1d558e73067359cd06b9313363cc9dc764a2c551611ac4R360). The `RetryOnErrorInterceptor` is now always installed by default. 

`ApolloClient.Builder` now takes a `retryOnError((ApolloRequest) -> Boolean)` to set a default `retryOnError` when not specified. By default it is false. You can see it in action  [here](https://github.com/apollographql/apollo-kotlin/pull/5690/files#diff-956eb6b81f65aa82149cac71ee464043a3c5b605f8bbe886417469241f3b48b9L28)

Reply:
so retrying subscriptions is no longer the default, it needs to be explicit, right?

Reply:
It never has been. In that last [pr](https://github.com/apollographql/apollo-kotlin/pull/5690/files#diff-aa6dd8789c20afff5b1d558e73067359cd06b9313363cc9dc764a2c551611ac4R360), the `RetryOnErrorInterceptor` is always used and opt-in is made with `ApolloClient.Builder.retryOnError {}`while in previous attempts, opt-in was made with `ApolloClient.Builder.retryOnErrorInterceptor()`

Reply:
All right - SGTM 👍